Add the ability to add a secret to an application

jeff-phillips-18 opened the pull request:
![image](https://user-images.githubusercontent.com/11633780/30130232-832435e0-9316-11e7-9465-ec0799e5b755.png)

@beanh66

spadgett commented on app/scripts/directives/addSecretToApplication.js:
```diff
@@ -0,0 +1,171 @@
+        statefulSets = _.toArray(statefulSetData.by('metadata.name'));
+        sortApplications();
+      });
+    };
```
We're doing this now in enough places now that it would be good to put this logic into a separate service (ApplicationsService?).

Not for this PR, though.

jeff-phillips-18 replied:
Heh, I thought the same and started working on it but I ran out of time to make the change.

Added issue: https://github.com/openshift/origin-web-common/issues/175

rhamilto commented on app/styles/_secrets.less:
```diff
@@ -98,3 +98,51 @@ dl.secret-data {
   }
 }
+
+.add-secret-to-application {
```
Nit:  alpha.  This rule should come first.

spadgett commented on app/scripts/directives/addSecretToApplication.js:
```diff
@@ -0,0 +1,171 @@
+        };
+
+        // For each container, add the new volume mount.
+        _.each(podTemplate.spec.containers, function(container) {
```
We should open a follow on for letting the user pick containers if there's more than one.

jeff-phillips-18 replied:
Added https://github.com/openshift/origin-web-console/issues/2024

rhamilto commented on app/styles/_secrets.less:
```diff
@@ -98,3 +98,51 @@ dl.secret-data {
+      }
+    }
+    .form-group > label {
+      font-size: 14px;
```
Use variable (e.g., @font-size-base + 1) so font sizes stay customizable.

spadgett commented on app/scripts/directives/addSecretToApplication.js:
```diff
@@ -0,0 +1,171 @@
+        // For each container, add the new volume mount.
+        _.each(podTemplate.spec.containers, function(container) {
+          container.envFrom = container.envFrom || [];
+          container.envFrom.push(newEnvFrom);
```
We should call `angular.copy` before we start changing the resource. Otherwise it mutates a value that's in the DataService cache

spadgett commented on app/scripts/directives/addSecretToApplication.js:
```diff
@@ -0,0 +1,171 @@
+      if (ctrl.addType === 'env') {
+        var newEnvFrom = {
+          secretRef: ctrl.secret.metadata
+        };
```
I'd prefer only adding valid API fields, even if other metdata is ignored. So

```js
var newEnvFrom = {
  secretRef: {
    name: ctrl.secret.metadata.name
  }
};
```

rhamilto commented on app/views/directives/add-secret-to-application.html:
```diff
@@ -0,0 +1,90 @@
+    <form name="addToApplicationForm" novalidate>
+      <fieldset ng-disabled="disableInputs">
+        <div class="form-group">
+          <label>Add this secret to application:</label>
```
Ideally label elements get a for attribute to associate them with the input they are for (here and throughout the file).

spadgett commented on app/scripts/directives/addSecretToApplication.js:
```diff
@@ -0,0 +1,171 @@
+        });
+      } else {
+        var generateName = $filter('generateName');
+        var name = generateName('volume-');
```
It might be better to use the secret name as the generateName prefix so it's more obvious what the volume is for.

rhamilto commented on app/views/directives/add-secret-to-application.html:
```diff
@@ -0,0 +1,90 @@
+          </div>
+        </div>
+        <div class="form-group">
+          <label>Add secret as:</label>
```
Semantically, this would be better as a legend element inside of a fieldset.  Not sure we have styling for that...

spadgett commented on app/views/browse/secret.html:
```diff
@@ -32,6 +38,8 @@ <h1 class="contains-actions">
               </li>
             </ul>
           </div>
+          <div class="pull-right" ng-hide="!('secrets' | canIDoAny)">
```
We should probably just remove the `canIDoAny` check if "Add to Application" is always enabled. Otherwise we need to update the `canIDoAny` filter for secrets.

jeff-phillips-18 replied:
Sorry this was a left over empty div.

spadgett commented on app/scripts/directives/addSecretToApplication.js:
```diff
@@ -0,0 +1,171 @@
+          NotificationsService.addNotification({
+            type: "success",
+            message: "Successfully added " + sourceKind + " " + ctrl.secret.metadata.name + " to " + targetKind + " " + ctrl.application.metadata.name + "."
+          });
```
Might be nice to link to the deployment config here.

jeff-phillips-18 commented:
@rhamilto Addressed your issues.

Also noticed the placement of the volume input did not match the design, updated:

![image](https://user-images.githubusercontent.com/11633780/30131963-babc253e-931c-11e7-8cd2-09ecb67532fc.png)

jeff-phillips-18 commented:
@spadgett I believe I have addressed your issues. Thanks!

spadgett commented on app/scripts/directives/addSecretToApplication.js:
```diff
@@ -0,0 +1,180 @@
+            message: "Successfully added " + sourceKind + " " + ctrl.secret.metadata.name + " to " + targetKind + " " + applicationToUpdate.metadata.name + ".",
+            links: [{
+              href: Navigate.resourceURL(applicationToUpdate),
+              label: "View " + targetKind
```
Nit: better to call `humanizeKind(applicationToUpdate.kind, true)` to get title case. This is consistent with our other notification links.

jeff-phillips-18 replied:
Fixed

spadgett commented:
[merge]
